wallet: prevent bugs from invalid transaction heights with asserts, comments, and refactoring

ryanofsky opened the pull request:
Originally, this PR fixed a wallet migration bug that could cause the watchonly wallet created by legacy wallet migration to have incorrect transaction height values. A different fix for the bug was implemented in #28609, but that PR did not add any test coverage that would have caught the bug, and didn't include other changes from this PR intended to prevent problems from invalid transaction heights.

This PR adds new asserts to catch invalid transaction heights, which would trigger test failures without bugfix in #28609. This PR also refactors code and adds comments to clarify assumptions and make it less likely a bug from invalid transaction height values would be introduced.

ryanofsky committed:
wallet, refactor: Add CWalletTx::updateState function

No change in behavior, this just moves code which updates transaction state to
a new method so it can be used after offline processes such as wallet
migration.

ryanofsky committed:
wallet: Add asserts to detect unset transaction height values

Also document GetTxDepthInMainChain preconditions better

ryanofsky commented:
Updated 3bb306142b7a66c8727bd19e8f0fb72c582bcdc2 -> e03b0fe72a8b1034e5221c8aed6fb31ddd056603 ([`pr/mig.1`](https://github.com/ryanofsky/bitcoin/commits/pr/mig.1) -> [`pr/mig.2`](https://github.com/ryanofsky/bitcoin/commits/pr/mig.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mig.1..pr/mig.2)) just adding code comment

achow101 commented:
Hmm, I wonder if we should instead make both the watchonly and solvables wallets with empty contexts (no chain), and then reload them when migration is done? That would also solve this problem.

ryanofsky commented:
> Hmm, I wonder if we should instead make both the watchonly and solvables wallets with empty contexts (no chain), and then reload them when migration is done? That would also solve this problem.

Yes, that would be less fragile than the current approach. Also I'm not sure why the main wallet needs to be created with an empty context anyway. It seems like the best solution would just be to create all the wallets normally and not need to reload any of them.

I do think the other changes in this PR make sense no matter how this issue is fixed, though. And I like that the fix implemented in this PR (0648dbbe092169b488e4db0d385aaa9d2726ddfe) is just a one line change.

ryanofsky commented:
Updated e03b0fe72a8b1034e5221c8aed6fb31ddd056603 -> dbed701d75a4a55cb34d797cfbae6910075ebec0 ([`pr/mig.2`](https://github.com/ryanofsky/bitcoin/commits/pr/mig.2) -> [`pr/mig.3`](https://github.com/ryanofsky/bitcoin/commits/pr/mig.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mig.2..pr/mig.3)) fixing a comment typo

achow101 commented:
> Yes, that would be less fragile than the current approach. Also I'm not sure why the main wallet needs to be created with an empty context anyway. It seems like the best solution would just be to create all the wallets normally and not need to reload any of them.

It has an empty context so that there isn't any possibility for an inconsistent state if a block and transaction is found.

ryanofsky commented:
> It has an empty context so that there isn't any possibility for an inconsistent state if a block and transaction is found.

I see, so a way to prevent the wallet from changing while it is being migrated.

In that case, I think building all the wallets offline and then reloading them like you suggested does make sense. Would be happy to review a PR if you have an idea of how to implement that. (Feel free to take any changes from this PR if they'd be useful. Or we could fix the bug directly with this PR and then make the change the loading in a later PR.)

achow101 commented:
ACK dbed701d75a4a55cb34d797cfbae6910075ebec0

furszy reviewed (commented):
> > It has an empty context so that there isn't any possibility for an inconsistent state if a block and transaction is found.
> 
> I see, so a way to prevent the wallet from changing while it is being migrated.
> 
> In that case, I think building all the wallets offline and then reloading them like you suggested does make sense. Would be happy to review a PR if you have an idea of how to implement that. (Feel free to take any changes from this PR if they'd be useful. Or we could fix the bug directly with this PR and then make the change the loading in a later PR.)

+1.
Not seeing a reason to create the watch-only and solvables wallets with a chain context then. They could also enter in an inconsistent state if/when a block reorg or transaction is found during migration.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3908,6 +3894,15 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
+                    // updateState will look up missing height values so the
+                    // watchonly wallet can function correctly after the
+                    // migration without needing to be reloaded.
+                    wtx->updateState(data.watchonly_wallet->chain());
                     // Add to watchonly wallet
                     if (!data.watchonly_wallet->AddToWallet(wtx->tx, wtx->m_state)) {
```
not really for this PR but shouldn't we also transfer the other wtx members too?
Like `mapValue` (for the "replaced_by_txid" and "replaces_txid"), `nTimeReceived`, `nTimeSmart`, `fFromMe`.

furszy replied:
nvm, just saw #28609 implementing it.

furszy commented:
I believe we can move to #28609, it fixes this issue and others in a more comprehensive manner.

achow101 commented:
Removing from the 26.0 milestone in favor of #28609.

ryanofsky commented:
Rebased dbed701d75a4a55cb34d797cfbae6910075ebec0 -> f06016d77d848133fc6568f287bb86b644c9fa69 ([`pr/mig.3`](https://github.com/ryanofsky/bitcoin/commits/pr/mig.3) -> [`pr/mig.4`](https://github.com/ryanofsky/bitcoin/commits/pr/mig.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mig.3-rebase..pr/mig.4)) due to conflict with #28609.

Also updated title and PR description since the original bug was fixed #28609, and now this PR is just adding asserts and cleanup to prevent related bugs.

furszy reviewed (commented):
Do you think the first commit is still necessary? I'm not seeing a reason for doing it when no other piece of code will use it.

ryanofsky commented:
> Do you think the first commit is still necessary? I'm not seeing a reason for doing it when no other piece of code will use it.

I think the first commit is useful to make it clear that some of the CWalletTx fields aren't serialized in the wallet and need to be loaded separately from the blocks database, and to have one function responsible for loading them. Otherwise the fields are just set conditionally in LoadToWallet away from the rest of transaction state code. The new function could also be useful in the future if we want to add more fields to transaction objects to improve conflict tracking, or if we want to add features that avoid keeping all transactions in memory. It's also a move-only change that should be easy to review.

furszy reviewed (commented):
> I think the first commit is useful to make it clear that some of the CWalletTx fields aren't serialized in the wallet and need to be loaded separately from the blocks database, and to have one function responsible for loading them.

Wouldn't be clearer to move the non-serialized members to optional?

But other than that, what make some noise on me about the approach is that we would be "hiding" a `cs_main` locking cause. Because `chain.findBlock()` locks the mutex internally.
Which, I know that it is currently not really important, but maybe in the future we would like to batch calls that lock `cs_main` as much as possible in the wallet.

Still, isn't that I'm against the refactoring, just would like to discuss alternatives.

ryanofsky commented:
> Wouldn't be clearer to move the non-serialized members to optional?
> 
> But other than that, what make some noise on me about the approach is that we would be "hiding" a `cs_main` locking cause. Because `chain.findBlock()` locks the mutex internally. Which, I know that it is currently not really important, but maybe in the future we would like to batch calls that lock `cs_main` as much as possible in the wallet.

Very much agree it makes sense to make locking and lookups explicit, and this change helps that, doesn't hurt it. It moves lookups to a function that explicitly is meant to do lookups, and explicitly takes a Chain parameter and doesn't have access to other wallet state. (CWalletTx used to have a CWallet* backpointer, but this was removed in b11a195ef450bd138aa03204a5e74fdd3ddced26 to force this explicitness.) The current block lookup code before this PR is in the middle of a `CWallet::LoadToWallet` function which is messier and is able to do the lookup because it has full access to the monolithic CWallet class.

Taking your batching example, if you wanted to look up the block heights as a batch, instead of between loading each transaction, this change gets closer to that goal, because it moves the lookup code out of the per-transaction LoadToWallet to function to a separate function exactly like you would do if you were making that change.

On your first idea of making heights use std::optional instead of -1 when they are uninitialized, that would be a fine change but it would not make code more explicit or improve code organization, so I see it as being more shallow. It would be complementary though.

furszy commented:
> (CWalletTx used to have a CWallet* backpointer, but this was removed in https://github.com/bitcoin/bitcoin/commit/b11a195ef450bd138aa03204a5e74fdd3ddced26 to force this explicitness.)

Thanks for that. Looks ugly.

> Taking your batching example, if you wanted to look up the block heights as a batch, instead of between loading each transaction, this change gets closer to that goal, because it moves the lookup code out of the per-transaction LoadToWallet to function to a separate function exactly like you would do if you were making that change.

I'm not sure if I would place it inside `transaction.h`. I would rather keep that file exclusively for primitive data.
This way, we can avoid providing access to the chain interface class in places that shouldn't have access to it, such as the `walletdb.cpp` loading process. The function wouldn't be member of `CWalletTx`; It would probably take the chain + a vector of wtx references for updating.

Sjors reviewed (approved):
utACK f06016d77d848133fc6568f287bb86b644c9fa69

furszy reviewed (commented):
Code review ACK f06016d

Have to admit that I'm still not sure about the first commit, but it isn't blocking if others are happy with it.

achow101 commented:
ACK f06016d77d848133fc6568f287bb86b644c9fa69
